api/src/main/java/marquez/service/SourceService.java

```diff
@@ -17,96 +17,70 @@
-
-        sources.inc();
-      }
-      return get(name).get();
-    } catch (UnableToExecuteStatementException e) {
-      log.error("Failed to create or update source '{}' with meta: {}", name.getValue(), meta, e);
-      throw new MarquezServiceException();
-    }
+  public Source createOrUpdate(@NonNull SourceName name, @NonNull SourceMeta meta) {
+    final Instant now = Instant.now();
+    final SourceRow row =
+        dao.upsert(
+            UUID.randomUUID(),
+            meta.getType().getValue(),
+            now,
+            now,
+            name.getValue(),
+            meta.getConnectionUrl().map(URI::toASCIIString).orElse(null),
+            meta.getDescription().orElse(null));
+    log.info("Added source '{}' with meta: {}", name.getValue(), meta);
+    return toSource(row);
   }
 
   public boolean exists(@NonNull SourceName name) throws MarquezServiceException {
-    try {
-      return dao.exists(name.getValue());
-    } catch (UnableToExecuteStatementException e) {
-      log.error("Failed to check for source '{}'.", name.getValue(), e);
-      throw new MarquezServiceException();
-    }
+    return dao.exists(name.getValue());
   }
 
-  public Optional<Source> get(@NonNull SourceName name) throws MarquezServiceException {
-    try {
-      return dao.findBy(name.getValue()).map(SourceService::toSource);
-    } catch (UnableToExecuteStatementException e) {
-      log.error("Failed to get source '{}'.", name.getValue(), e);
-      throw new MarquezServiceException();
-    }
+  public Optional<Source> get(@NonNull SourceName name) {
+    return dao.findBy(name.getValue()).map(SourceService::toSource);
   }
 
-  public ImmutableList<Source> getAll(int limit, int offset) throws MarquezServiceException {
+  public ImmutableList<Source> getAll(int limit, int offset) {
     checkArgument(limit >= 0, "limit must be >= 0");
     checkArgument(offset >= 0, "offset must be >= 0");
-    try {
-      final ImmutableList.Builder<Source> sources = ImmutableList.builder();
-      final List<SourceRow> rows = dao.findAll(limit, offset);
-      for (final SourceRow row : rows) {
-        sources.add(toSource(row));
-      }
-      return sources.build();
-    } catch (UnableToExecuteStatementException e) {
-      log.error("Failed to get sources.", e);
-      throw new MarquezServiceException();
+    final ImmutableList.Builder<Source> sources = ImmutableList.builder();
+    final List<SourceRow> rows = dao.findAll(limit, offset);
+    for (final SourceRow row : rows) {
+      sources.add(toSource(row));
     }
+    return sources.build();
   }
 
   static Source toSource(@NonNull final SourceRow row) {
     return new Source(
-        SourceType.valueOf(row.getType()),
+        SourceType.of(row.getType()),
         SourceName.of(row.getName()),
         row.getCreatedAt(),
         row.getUpdatedAt(),
-        URI.create(row.getConnectionUrl()),
+        row.getConnectionUrl().map(URI::create).orElse(null),
```

Comment:
why unbox optionals?

Reply:
This can lead to sneaky exceptions on GET operations which can be an issue if someone wanted to integrate with marquez directly